extensions/oidc-devservices-keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/KeycloakDevServicesProcessor.java

```diff
@@ -0,0 +1,395 @@
+import io.quarkus.deployment.builditem.LaunchModeBuildItem;
+import io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem;
+import io.quarkus.deployment.builditem.ServiceStartBuildItem;
+import io.quarkus.runtime.LaunchMode;
+import io.quarkus.runtime.configuration.ConfigUtils;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.mutiny.core.buffer.Buffer;
+import io.vertx.mutiny.ext.web.client.WebClient;
+
+public class KeycloakDevServicesProcessor {
+    private static final Logger LOG = Logger.getLogger(KeycloakDevServicesProcessor.class);
+
+    private static final String CONFIG_PREFIX = "quarkus.oidc.";
+    private static final String AUTH_SERVER_URL_CONFIG_KEY = CONFIG_PREFIX + "auth-server-url";
+    private static final String APPLICATION_TYPE_CONFIG_KEY = CONFIG_PREFIX + "application-type";
+    private static final String CLIENT_ID_CONFIG_KEY = CONFIG_PREFIX + "client-id";
+    private static final String CLIENT_SECRET_CONFIG_KEY = CONFIG_PREFIX + "credentials.secret";
+    private static final String KEYCLOAK_URL_KEY = "keycloak.url";
+
+    private static final int KEYCLOAK_EXPOSED_PORT = 8080;
+    private static final String KEYCLOAK_DOCKER_REALM_PATH = "/tmp/realm.json";
+    private static final String KEYCLOAK_USER_PROP = "KEYCLOAK_USER";
+    private static final String KEYCLOAK_PASSWORD_PROP = "KEYCLOAK_PASSWORD";
+    private static final String KEYCLOAK_VENDOR_PROP = "DB_VENDOR";
+    private static final String KEYCLOAK_IMPORT_PROP = "KEYCLOAK_IMPORT";
+    private static final String KEYCLOAK_ADMIN_USER = "admin";
+    private static final String KEYCLOAK_ADMIN_PASSWORD = "admin";
+    private static final String KEYCLOAK_DB_VENDOR = "H2";
+    private static volatile List<Closeable> closeables;
+    private static volatile DevServicesConfig capturedDevServicesConfiguration;
+    private static volatile boolean first = true;
+    private final IsDockerWorking isDockerWorking = new IsDockerWorking(true);
+
+    KeycloakBuildTimeConfig config;
+
+    @BuildStep(onlyIfNot = IsNormal.class)
+    public DevServicesConfigBuildItem startOIDCContainer(LaunchModeBuildItem launchMode,
+            BuildProducer<RunTimeConfigurationDefaultBuildItem> runTimeConfiguration,
+            BuildProducer<ServiceStartBuildItem> serviceStartBuildItemBuildProducer,
+            BuildProducer<DevServicesNativeConfigResultBuildItem> devServices) {
+
+        DevServicesConfig currentDevServicesConfiguration = config.devservices;
+        // figure out if we need to shut down and restart any existing Keycloak container
+        // if not and the Keycloak container have already started we just return
+        if (closeables != null) {
+            boolean restartRequired = launchMode.getLaunchMode() == LaunchMode.TEST;
+            if (!restartRequired) {
+                restartRequired = !currentDevServicesConfiguration.equals(capturedDevServicesConfiguration);
```

Comment:
We should also check the realm file for changes, so if the user modifies the configuration this should be taken into account.

Reply:
Hi Stuart, @stuartwdouglas - I've been thinking about it and nearly implemented a size check, but then realized that it is not really needed - if the realm file exists then Keycloak will be initialized from it - but Dev UI always works against the live current Keycloak instance - i.e, if the realm has been updated via KC Admin Console and re-exported into the same file which KC was initialized from - it won't change anything since whatever updates that have been applied to this realm are live so DevUI will pick them up indirectly when requesting the tokens or letting the users to sign in to KC.

Reply:
Remember there are two KC instances, the test one and the dev one. Even if you assume it is only updated by export and not by manual modification the test one would still need to be updated to be in sync with the dev one. 

Also the dev and test ones may have separate files (as you could set a different file name in a %test profile), so this still should check this.

Reply:
Hi @stuartwdouglas `manual` modification of Keycloak realm is most likely (if not definitely) is not an option - it can be up to 80K and it is quite a complex JSON - I doubt anyone is attempting to change it manually so I would not be concerned about missing the manual updates which might affect a given KC instance - I think we can safely assume/require that the realm file modifications should only be done from the KC admin console's `export realm` option. Myself and others have tried to trim `realm.json` a few times simply to make it smaller - but this is not done to affect the behavior of KC as such :-), it is just a trial and error attempt to produce a small size file which is easier to read/etc...

Re the test case - this one is easy, `quarkus-oidc` does not support KC specific properties or are you saying that we can have 2 sets of `keycloak.devservices` properties - one for starting a devservices container and another one (in a test profile) for running the tests ? Should we allow it ? (I'd be a bit confused if it were the case).

I think as far as the file check is concerned, the only check which can work is a file size check - that is easy enough to do. I'd be concerned that the devservice users will see the container restarting immediately after exporting the realm from the current KC instance console - which would make the current browser view stale: ex, the user is working with `localhost:12345/auth/realms/admin`, exports the updated realm, continues clicking around in that console, while Quarkus has already stopped the container which backs it up and now it runs at `localhost:67890/auth/realms/admin`

Reply:
@stuartwdouglas update, `RealmRepresentation` from `keycloak-core` has no `equals` so if if we were to check for the possible updates it would have to be a file size check. I'd prefer though to wait with it until we see an actual issue with not checking it :-), see above for some concerns as well

Reply:
Of course we can have different properties for running the tests. If you have a `%test.quarkus.oidc.devservices.realm-path` you can use a different file for tests. We don't need a .equals() check for modification all you need to check is if the last modified time has changed. 

Quarkus won't actually restart until you hit the app with the browser. If you were really worried about it you could make the restarted container use the same port.

Reply:
@stuartwdouglas Sure, the realm can either be a classpath resource (possibly bundled in a jar) or a file system resource, for the  latter case I can add this check. Good point about the same port in case of restarting the container - it should probably be done anyway - but I'll look at it if there will indeed be some issues related to syncing the views with the running containers

Re running the tests with the alternative devservices setup - not sure it really works in practice though - i can certainly see how I can start building tests while I'm doing the dev against the current KC container - but eventually I'd have tests depending on the test resource factory starting a container (as is the case for all the OIDC tests).. That said, I see, the dev mode tests can indeed depend now on the devservices config... interesting, may be I will try to update Quarkus OIDC dev mode test to try once the code is merged...